File: app/src/main/java/dev/yiray/qickmvivm/ui/home/HomeViewModel.java
package dev.yiray.qickmvivm.ui.home;

import android.util.Log;

import dev.yiray.qickmvivm.base.BaseState;
import dev.yiray.qickmvivm.base.BaseViewModel;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Observable;

public class HomeViewModel extends BaseViewModel<HomeViewState, HomeView.Action> {
    private TaskInteractor interactor;

    private HomeViewState memorized;

    public HomeViewModel() {
        interactor = new TaskInteractor();
    }

    @Override
    protected void bindIntent(HomeView.Action intent) {
        Observable<HomeViewState> observableInput = intent.observableInput()
                .switchMap(charSequence -> Observable.just(new HomeViewState.InputState(charSequence.toString())));
        Observable<HomeViewState> observableCheckBox = intent.observableCheckBox()
                .map(HomeViewState.CheckBoxState::new);
        Observable<HomeViewState> observableNext = intent.observableNextPage()
                .map(aBoolean -> new HomeViewState.NextPage());
        Observable<HomeViewState> observableListUpdated = intent.observableUpdated()
                .map(updatedTask -> new HomeViewState.ListModified(updatedTask.getPosition(), updatedTask.getTask()));

        Observable<HomeViewState> intents = Observable.merge(
                observableInput,
                observableCheckBox,
                observableNext,
                observableListUpdated
        ).observeOn(AndroidSchedulers.mainThread());

        HomeViewState state;

        if (memorized == null) {
            state =  new HomeViewState.ViewState.Builder().build();
        } else {
            state = memorized;
        }

        Observable<HomeViewState> stateObservable =
                intents.scan(state, this::reducer);

        subscribe(stateObservable);
    }

    private HomeViewState reducer(HomeViewState previousState, HomeViewState newState) {
            Log.i("reducer", newState.toString());
            Log.i("previousState", previousState.toString());

            if (newState instanceof BaseState.SideEffect) {
                handleSideEffects(previousState, newState);
            }

            HomeViewState state = newState;

            if (newState instanceof HomeViewState.ListModified) {
                state = interactor.updateTaskList(
                        previousState.toViewState(),
                        (HomeViewState.ListModified) newState
                );
            }

            memorized = previousState.toViewState().copy(state.toViewState());

            return memorized;
    }
}

File: app/src/main/java/dev/yiray/livedatamvi/ui/home/HomeViewModel.kt
package dev.yiray.livedatamvi.ui.home

import dev.yiray.livedatamvi.ui.home.HomeView.UpdateTask
import dev.yiray.livedatamvi.ui.home.HomeViewState.NextPage
import android.util.Log
import dev.yiray.livedatamvi.base.BaseViewModel
import dev.yiray.livedatamvi.ui.home.HomeViewState.InputState
import dev.yiray.livedatamvi.ui.home.HomeViewState.ListModified
import dev.yiray.livedatamvi.base.BaseState.SideEffect
import kotlinx.coroutines.flow.*

class HomeViewModel : BaseViewModel<HomeViewState, HomeView.Action>() {
    private val interactor: TaskInteractor = TaskInteractor()
    private var memorized: HomeViewState? = null
    public override fun bindIntent(intent: HomeView.Action) {
        val observableInput = intent.observableInput()
            .drop(1)
            .map { charSequence: CharSequence ->
                InputState(
                    charSequence.toString()
                )
            }

        val observableCheckBox = intent.observableCheckBox()
            .map { isChecked: Boolean? ->
                HomeViewState.CheckBoxState(
                    isChecked!!
                )
            }
        val observableNext = intent.observableNextPage()
            .map { NextPage() }
        val observableListUpdated = intent.observableUpdated()
            ?.map { updatedTask: UpdateTask? ->
                ListModified(
                    updatedTask!!.position,
                    updatedTask.task
                )
            }

        val intents = merge(
            observableInput,
            observableCheckBox,
            observableNext,
            observableListUpdated!!
        )

        val state: HomeViewState? = if (memorized == null) {
            HomeViewState.ViewState()
        } else {
            memorized
        }
        val stateObservable = intents.scan(
            state!!,
            { previousState: HomeViewState?, newState: HomeViewState ->
                reducer(
                    previousState,
                    newState
                )
            })
        subscribe(stateObservable)
    }

    private fun reducer(previousState: HomeViewState?, newState: HomeViewState): HomeViewState {
        Log.i("reducer", newState.toString())
        Log.i("previousState", previousState.toString())
        var state: HomeViewState? = newState

        when (newState) {
            is SideEffect -> {
                if (previousState != null) {
                    handleSideEffects(previousState, newState)
                }
            }
            is ListModified -> {
                state = interactor.updateTaskList(
                    previousState!!.toViewState(),
                    newState
                )
            }
        }

        memorized = previousState!!.toViewState()?.copy(state!!.toViewState())
        return memorized!!
    }

}